lib/ShareProvider.js

```diff
@@ -0,0 +1,466 @@
+        }
+        // if this is a share of a selected item, get any microformats
+        if (!pageData.microformats && target) {
+          messageManager.addMessageListener("PageMetadata:MicroformatsResult", _dataFn = msg => {
+            messageManager.removeMessageListener("PageMetadata:MicroformatsResult", _dataFn);
+            pageData.microformats = msg.data;
+            ShareUtils.sharePage(providerOrigin, pageData, target);
+          });
+          gBrowser.selectedBrowser.messageManager.sendAsyncMessage("PageMetadata:GetMicroformats", null, {target});
+          return;
+        }
+
+        let provider = Social._getProviderFromOrigin(providerOrigin);
+        if (!provider || !provider.shareURL) {
+          return;
+        }
+
+        let shareEndpoint = OpenGraphBuilder.generateEndpointURL(provider.shareURL, pageData);
+        window.open(shareEndpoint, "share-dialog", "chrome");
+
+        eventTracker.handleUserEvent({
+          event: "SHARE_FROM_TOOLBAR",
+          provider: provider.origin
+        });
+      }
+    }
+  };
+}
+
+// This is to register a ShareUtils object in the browser chrome window and
+// to hide the existing social share (paper airplane) button.
+const Overlay = {
+  init: eventTracker => {
+    for (let win of CustomizableUI.windows) {
+      Overlay.setWindowScripts(win, eventTracker);
+    }
+    Services.obs.addObserver(Overlay, "browser-delayed-startup-finished", false);
+  },
+  uninit: () => {
+    Services.obs.removeObserver(Overlay, "browser-delayed-startup-finished");
+    for (let win of CustomizableUI.windows) {
+      delete win.ShareUtils;
+      if (win.SocialShare.shareButton) {
+        win.SocialShare.shareButton.removeAttribute("hidden");
+      }
+    }
+  },
+  observe: window => {
+    Overlay.setWindowScripts(window);
+  },
+  setWindowScripts: (window, eventTracker) => {
+    Object.defineProperty(window, "ShareUtils", windowProperty(window, eventTracker));
+
+    if (window.SocialShare.shareButton) {
+      window.SocialShare.shareButton.setAttribute("hidden", "true");
+    }
+  }
+};
+
+function createElementWithAttrs(document, type, attrs) {
+  let element = document.createElement(type);
+  Object.keys(attrs).forEach(attr => {
+    element.setAttribute(attr, attrs[attr]);
+  });
+  return element;
+}
+
+const DEFAULT_OPTIONS = {eventTracker: {handleUserEvent() {}}};
+
+function ShareProvider(options = {}) {
+  this.options = Object.assign({}, DEFAULT_OPTIONS, options);
+  this.eventTracker = this.options.eventTracker;
+}
+
+ShareProvider.prototype = {
+
+  /**
+   * Initialize Share Provider
+   */
+  init: Task.async(function*() {
+    yield this._setupProviders();
+    this._createButton();
+    Overlay.init(this.eventTracker);
+  }),
+
+  /**
+   * Set up the prefs and enable a default set of social providers if the user
+   * hasn't enabled any.
+   */
+  _setupProviders: Task.async(function*() {
+    if (!Services.prefs.prefHasUserValue("social.activeProviders")) {
```

Comment:
This also needs a check that there are not existing providers already (to prevent changing providers for existing share users).  && !SocialService.hasEnabledProviders